incubator/binding-openapi/src/main/java/io/aklivity/zilla/runtime/binding/openapi/internal/config/OpenapiOptionsConfigAdapter.java

```diff
@@ -163,14 +175,88 @@ private OpenApi parseOpenApi(
         String openapiText)
     {
         OpenApi openApi = null;
-        try (Jsonb jsonb = JsonbBuilder.create())
+        if (validateOpenapiSchema(openapiText))
         {
-            openApi = jsonb.fromJson(openapiText, OpenApi.class);
+            try (Jsonb jsonb = JsonbBuilder.create())
+            {
+                openApi = jsonb.fromJson(openapiText, OpenApi.class);
+            }
+            catch (Exception ex)
+            {
+                rethrowUnchecked(ex);
+            }
+        }
+        return openApi;
+    }
+
+    private boolean validateOpenapiSchema(
+        String openapiText)
+    {
+        List<Exception> errors = new LinkedList<>();
+
+        boolean valid = false;
+
+        try
+        {
+            JsonValidationService service = JsonValidationService.newInstance();
+
+            String openApiVersion = detectOpenApiVersion(openapiText);
+            InputStream schemaInput = selectSchemaPathForVersion(openApiVersion);
+
+            JsonSchema schema = service.readSchema(schemaInput);
+            ProblemHandler handler = service.createProblemPrinter(msg -> errors.add(new ConfigException(msg)));
+
+            String readable = openapiText.stripTrailing();
+            Reader openapiReader = new StringReader(readable);
+
+            JsonReader reader = service.createReader(openapiReader, schema, handler);
+
+            JsonStructure json = reader.read();
+            valid = json != null;
```

Comment:
Please use `JsonProvider provider = service.createJsonProvider(schema, problemHandler)` to create a validating `JsonProvider`.
Then use `JsonbBuilder.newBuilder().withProvider(provider).build()` instead of `Jsonb.create()`.

This will let you combine both methods into one and do a single pass of the the openapi document during (validating) parse / build.